src/coreclr/tools/Common/Compiler/ObjectWriter/WasmObjectWriter.cs

```diff
@@ -0,0 +1,233 @@
+    /// </summary>
+    internal sealed class WasmObjectWriter : ObjectWriter
+    {
+        // must be stored in little-endian order
+        private const uint WasmMagicNumber = 0x6d736100;
+
+        // must be stored in little-endian order
+        private const uint WasmVersion = 0x1;
+
+        private WasmAbiContext _wasmAbiContext;
+
+        public WasmObjectWriter(NodeFactory factory, ObjectWritingOptions options, WasmAbiContext ctx, OutputInfoBuilder outputInfoBuilder)
+            : base(factory, options, outputInfoBuilder)
+        {
+            _wasmAbiContext = ctx;
+        }
+
+        private void EmitWasmHeader(Stream outputFileStream)
+        {
+            SectionData data = new();
+            SectionWriter headerWriter = new(this, 0, data);
+            headerWriter.WriteLittleEndian(WasmMagicNumber);
+            headerWriter.WriteLittleEndian(WasmVersion);
+
+            data.GetReadStream().CopyTo(outputFileStream);
+        }
+
+        public override void EmitObject(Stream outputFileStream, IReadOnlyCollection<DependencyNode> nodes, IObjectDumper dumper, Logger logger)
+        {
+            ArrayBuilder<WasmFuncType> methodSignatures = new();
+            ArrayBuilder<IMethodBodyNode> methodBodies = new();
+            foreach (DependencyNode node in nodes)
+            {
+                if (node is IMethodBodyNode methodNode)
+                {
+                    methodSignatures.Add(_wasmAbiContext.GetSignature(methodNode.Method));
+                    methodBodies.Add(methodNode);
+                    // TODO: record relocations and attached data
+                }
+            }
+
+            var uniqueSignatures = methodSignatures.ToArray().Distinct();
+            var signatureMap = uniqueSignatures.Select((sig, i) => (sig, i)).ToDictionary();
+
+            EmitWasmHeader(outputFileStream);
+            EmitSection(() => WriteTypeSection(uniqueSignatures, logger), outputFileStream, logger);
+            EmitSection(() => WriteFunctionSection(methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
+            EmitSection(() => WriteExportSection(methodBodies.ToArray(), methodSignatures.ToArray(), signatureMap, logger), outputFileStream, logger);
+            EmitSection(() => WriteCodeSection(methodBodies.ToArray(), logger), outputFileStream, logger);
+        }
+
+        private WasmSection WriteExportSection(IReadOnlyList<IMethodBodyNode> methodNodes, IReadOnlyList<WasmFuncType> methodSignatures,
+            Dictionary<WasmFuncType, int> signatureMap, Logger logger)
+        {
+            WasmSection exportSection = new WasmSection(WasmSectionType.Export, new SectionData(), "export");
+            SectionWriter writer = exportSection.Writer;
+            // Write the number of exports
+            writer.WriteULEB128((ulong)methodNodes.Count);
+            for (int i = 0; i < methodNodes.Count; i++)
+            {
+                var methodNode = methodNodes[i];
```

Comment:
Nit: (here and elsewhere) we have some guidelines for `var` that optimize for code reading: https://github.com/dotnet/runtime/blob/b54c5d53ca8208e9bb5c95a2fbbd27f30df09b9d/docs/coding-guidelines/coding-style.md?plain=1#L22-L23. So these `var`s need to be changed.

(I personally just avoid using `var` everywhere in this project, since it is a simpler rule to follow)